### DISCUSSION TITLE
Interface design to show untracked files on-demand

### DISCUSSION THREAD

**Discussion by vermiculus** [magit/magit#4932] [Ideas]
Collecting untracked files in a large worktree can be a slow operation – so much so that Git will detect when it's slow and provide instructions for how to disable it (`status.showUntrackedFiles`).

Magit respects this configuration today in `magit-insert-untracked-files`. If `status.showUntrackedFiles=no`, the section is skipped. This speeds up the status buffer considerably at the cost of making it much harder to track new files (unless I'm missing some clever workflow – the best I've got is `: add <very-long-path> RET`).

I'd like to make it easier to show untracked files when needed without setting/resetting config. In my use-case, we have an exceptionally large repository (with an equally large worktree) where `git status -uno` is a practical necessity for every-day work. However, often enough, developers will create new files deep in the worktree and will want a way to quickly add them.

After a bit of iteration (including a `magit-status-config` transient that didn't really work out), the best I can really come up with is something like the following:

```
Head:     main deadbeef
Push:     origin/main deadbeef

Untracked files:
Skipped per "status.showUntrackedFiles". Press RET to list them now.

Unstaged changes (3)
modified   lisp/magit-git.el
modified   lisp/magit-status.el
modified   lisp/magit.el
```

where `RET` would pop open a dedicated buffer to stage untracked files – perhaps with some marking system (a la dired) to avoid extraneous refreshes.

While this would definitely work, this still feels like it can/should be improved. With this design's reliance on the separate buffer/mode, it's definitely a large enough effort where I'd like some amount of directional buy-in before developing a draft.

Some problems I identified in previous iterations (and the solutions I identified, in the hopes that there may be better solutions):

* There doesn't seem to be any good space in `magit-status-mode-map` left to show a configuration transient. Using the section's map resolves this.
* Making it easy to toggle the section on/off doesn't make it fast to add multiple different files. Staging each file with the existing section implementation triggers a refresh. Using a mark->execute workflow resolves this.

What do you think?

**Comment by tarsius** [comment]
Have you tried `magit-stage-file`?

That could be improved to offer the visited file as default and to optionally offer ignored files:

```elisp
(defun magit-stage-file (file &optional force)
  "Stage all changes to FILE.
With a prefix argument, always prompt for the file to stage.
Otherwise, if in a Magit buffer and there is a file at point,
which could be staged, stage without prompting.  When prompting,
offer the file at point or the visited file as default, if it
can be staged.  With two prefix arguments, allow staging ignored
files."
  (interactive
   (let* ((force (>= (prefix-numeric-value current-prefix-arg) 16))
          (atpoint (magit-section-value-if 'file))
          (current (magit-file-relative-name))
          (choices (nconc (magit-unstaged-files)
                          (magit-untracked-files force)))
          (default (car (member (or atpoint current) choices))))
     (list (if (or current-prefix-arg
                   (not atpoint)
                   (not default))
               (magit-completing-read "Stage file" choices
                                      nil t nil nil default)
             default)
           force)))
  (magit-with-toplevel
    (magit-stage-1 (and force "--force")
                   (list file))))
```

It could also be added to `magit-file-dispatch` to make it easy to invoke from a file visiting buffer.

**Reply by vermiculus** [reply]
I think this almost solves both main problems, with a little tweaking:

* How do I know which files are untracked?

    The above definition makes this easy with its call to `magit-untracked-files`.

* How can I stage multiple files at a time to limit refreshes?

    If this (or a derivative) could use `completing-read-multiple`, this would be addressed. I'll freely admit that I don't think I myself have quite mastered CRM as a user, but it's at least a standard emacs way of doing things and UI packages abound (a la vertico) that can potentially make it as intuitive as marking in a dired-like buffer.

What are your thoughts on teaching the *Untracked Files* section to show this when `-uno` is in effect? Seems a little odd at first, but if you really don't want the section, you could disable it entirely with `remove-hook` or similar.

I think either way, I've got enough so far to start working on a patch. I'll open up a WIP PR shortly.

**Reply by vermiculus** [reply]
Opened #4934. I'm a little surprised GitHub doesn't cross-link these conversations with the reference in the PR, but so it is.

**Reply by tarsius** [reply]
I have though more about my approach. Unfortunately that lead me down a path of "lets finally do all of this properly". Proceeding in that direction will be beneficial, but should not be rushed and involves quite a few design decisions.

So I have chosen to give you a heads up, instead of trying to have something ready *any time now*.

> I think this almost solves both main problems, with a little tweaking:
> 
>     * How do I know which files are untracked?
>       The above definition makes this easy with its call to `magit-untracked-files`.

So you are saying that `(nconc (magit-unstaged-files) (magit-untracked-files force))` is not good enough because it lists both kinds of files that could be staged in one set?

We could of course add a separate `magit-stage-untracked` command, or do something with prefix arguments, but I would rather avoid both.

>     * How can I stage multiple files at a time to limit refreshes?
>       If this (or a derivative) could use `completing-read-multiple`, this would be addressed.

I intend to do that.

>       I'll freely admit that I don't think I myself have quite mastered CRM as a user, but it's at least a standard emacs way of doing things and UI packages abound (a la vertico) that can potentially make it as intuitive as marking in a dired-like buffer.

Dired is one of the complications I hinted at above. I have to rething all the `magit-file-*` and `magit-*-file` commands taking Dired, and what not, into account. Some already do support dired, some don't. Some work only in file-visiting buffers, some only in Magit buffers, some also in Dired buffers. and/or/xor. And what about `magit-file-dispatch`, that seems a bit odd too. It's a bit complicated.

> What are your thoughts on teaching the _Untracked Files_ section to show this when `-uno` is in effect? Seems a little odd at first, but if you really don't want the section, you could disable it entirely with `remove-hook` or similar.

It definitely should not be the default. You could *enable* it with `add-hook` or similar. ;P

But I don't think it is needed at all anyway.

Your approach and mine have something in common: regardless of how we get there, some command eventually makes the user select one or more files from a list of untracked files.

The difference is that you want to do that by having the user go to the status buffer and then navigate to the "there might be untracked files" section and then press `s`; while I want to make a command available that can be invoked from anywhere, and bind it to some prefix sequence, say `C-c f , s` (`C-c f` (or the horrible `C-c M-g`) for `magit-file-dispatch`, followed by `, s`).

I think my approach is better because it doesn't force the user to go to a specific location to invoke this action, and because it doesn't require adding a section to the status buffer, which does not actually provide *any* information. (If we *do* show untracked files in the status buffer, then that obviously is information. But if we just display such a section, without either listing files, or at least saying how many files there are, then that doesn't tell us anything. If listing untracked is too slow, then counting untracked files, likely also is too slow.)

> I think either way, I've got enough so far to start working on a patch. I'll open up a WIP PR shortly.

**Reply by tarsius** [reply]
I have just pushed some changes that allows you to conveniently stage untracked files.

- In a file-visiting buffer `C-c M-g s` stages that file. If it is ignored, then the user has to confirm.
- Elsewhere `C-c M-g s` prompts for a file tracked or untracked (but not ignored) file, while `C-c M-g C-u s`.

The `M-g` is inconvenient. I recommend binding `C-c f` instead of `C-c M-g` to `magit-file-dispatch`.

Let me know if you still think a "there might be untracked files" section, as you originally proposed, is still needed.

**Reply by vermiculus** [reply]
>> * How do I know which files are untracked?
>> The above definition makes this easy with its call to `magit-untracked-files`.
>
> So you are saying that `(nconc (magit-unstaged-files) (magit-untracked-files force))` is not good enough because it lists both kinds of files that could be staged in one set?

If I can translate for my own past self, I was confirming that your snippet *did* answer that question (and how it did so). Likely for my own reference later.

> Your approach and mine have something in common: regardless of how we get there, some command eventually makes the user select one or more files from a list of untracked files.
>
> The difference is that you want to do that by having the user go to the status buffer and then navigate to the "there might be untracked files" section and then press s; while I want to make a command available that can be invoked from anywhere, […].
>
> I think my approach is better because it doesn't force the user to go to a specific location to invoke this action, and because it doesn't require adding a section to the status buffer, which does not actually provide any information. […]

I certainly have no argument against a command that can be invoked from anywhere :-) I believe the nugget I was attempting to hit with the status buffer section was simply discoverability. Not many repositories have a need to use `status -uno` – and our repositories need to set it by default for scalability reasons. That said,

> It definitely should not be the default. You could enable it with `add-hook` or similar. ;P

is absolutely true and likely what I should do in my distribution of Magit (+ other 'starter kit'-esque goodies).

I've yet to read your patch, but assuming it defines a command that can prompt for files on its own (rather than needing a `buffer-file-name` context), then I imagine that'll work just fine :-)

---

Thanks as always!

Aside, I have been continuing in my distribution efforts we discussed over email with moderate success (i.e., I haven't been laughed out for suggesting we recommend everyone use emacs). I may be reaching out for advice on making Windows more performant – as I'm sure you've heard about time and time again (especially considering the *Performance* section in the manual). I've got more investigation to do on my own, though.

**Reply by tarsius** [reply]
Time flies...

I have lost context, so on the technical side, I'll wait until you have taken my additions for a spin (once you have done that, I'll even have to read up on those myself :wink:).

> Aside, I have been continuing in my distribution efforts we discussed over email with moderate success (i.e., I haven't been laughed out for suggesting we recommend everyone use emacs).

:+1:

> I may be reaching out for advice on making Windows more performant – as I'm sure you've heard about time and time again (especially considering the Performance section in the manual). I've got more investigation to do on my own, though.

That's one thing that (I believe) hasn't made it into that section yet: https://github.blog/2022-06-29-improve-git-monorepo-performance-with-a-file-system-monitor/.

**Reply by tarsius** [reply]
Given the current behavior of `magit-stage-file`, I am closing this.
